hooks/useTokenAllowances.tsx

```diff
@@ -0,0 +1,45 @@
+import { TokenListItem } from "@/hooks/useTokenList";
+import { networks } from "@/utils/networks";
+import { getRailgunSmartWalletContractForNetwork } from "@railgun-community/quickstart";
+import { readContracts } from "@wagmi/core";
+import useSWR from "swr";
+import { useAccount, useNetwork, erc20ABI } from "wagmi";
+
+const useTokenAllowances = ({ tokenList }: { tokenList: TokenListItem[] }) => {
+  const { chain } = useNetwork();
+  const { address } = useAccount();
+  const network = networks[chain?.id || 1];
+
+  const chainId = chain?.id || 1; // default to mainnet if no chain id
+  const { isLoading, error, data } = useSWR(
+    `useTokenAllowances-${chainId}-${tokenList.length}`,
+    async () => {
+      if (!tokenList || tokenList.length === 0) {
+        return new Map();
+      }
+
+      const contractAddress = getRailgunSmartWalletContractForNetwork(
+        network.railgunNetworkName
+      ).address;
+      const readContractsArgs = tokenList.map((token) => {
+        return {
+          abi: erc20ABI,
+          functionName: "allowance",
+          address: token.address,
+          args: [address, contractAddress],
+        };
+      });
+      const data = await readContracts({
+        contracts: readContractsArgs,
+      });
+      const allowancesPerToken = new Map();
+      const tokenListWithUserBalance = tokenList.map((token, i) => {
```

Comment:
This `tokenListWithUserBalance` value is never used. I think this should be a `.forEach` instead of `.map` if we don't need the return value

But I'm wondering if the `useTokenAllowances` and `useTokenBalances` hooks should just be merged into a `useTokens` hook (or whatever name you like), since we'll always need both, and by merging them we can remove an RPC call

Reply:
Yes, I think that is a good idea. Do you like the use of a map instead of adding and extra value on a token item in the token list? FYI, I personally prefer the map

Reply:
> Do you like the use of a map instead of adding an extra value on the token item in the token list? FYI, I personally prefer the map

Sorry not sure I totally follow, do you mind expanding on the two options?

Reply:
Yes, the token list is an array of type TokenListItem, and we build that array by fetching the the token list and then adding the balance values to all the tokens in the token list.

For the allowances we fetch all the allowance values and then create a map with a key of token address and a value of allowance.

My question was if we are going merge both balances and allowances together which data structure would you prefer. And the structures would be as follows

1. An array of tokens from the token list with both the allowance and balance added to each token
2. A map with a key of token address and a value of an object with a key of balance and allowance.

I prefer 2 because it is more efficient to get a single value then 1. Is that clearer?

Reply:
Ahh yes I see. Thanks! I'm good with a `Map`, we can still get all keys from the map if needed. @wildmolasses let us know if you have any thoughts/objections here